### PR TITLE
Feat 565 remove redundant thrust dialect conditional

### PR DESCRIPTION
**Pull request by ZelboK**
Currently there is wide use of  `THRUST_CPP_DIALECT >= 2011` through out the codebase. This conditional always leads to true, so it can be safely removed.

### CODE REVIEW

File: thrust/examples/cuda/async_reduce.cu [hunk 1]
```diff
@@ -59,10 +57,7 @@ int main()
   // reset the result
   result[0] = 0;
 
-#if THRUST_CPP_DIALECT >= 2011
-  // method 2: use std::async to create asynchrony
```
Comment (ZelboK) [inline thrust/examples/cuda/async_reduce.cu:63]: undo comment removal

File: thrust/internal/benchmark/bench.cu [hunk 2]
```diff
@@ -712,7 +702,7 @@ struct shuffle_trial_base : trial_base<TrialKind>
     randomize(input);
   }
 };
-#endif
```
Comment (ZelboK) [inline thrust/internal/benchmark/bench.cu:715]: remove

File: thrust/thrust/detail/config/cpp_compatibility.h [hunk 3]
```diff
@@ -52,25 +52,18 @@
 //#  if   THRUST_CPP_DIALECT >= 2017
 //#    define THRUST_INLINE_CONSTANT                 inline constexpr
 //#    define THRUST_INLINE_INTEGRAL_MEMBER_CONSTANT inline constexpr
-#  if THRUST_CPP_DIALECT >= 2011
+// not that comfortable touching the
```
Comment (ZelboK) [inline thrust/thrust/detail/config/cpp_compatibility.h:55]: remove comment

File: thrust/thrust/system/cuda/detail/par.h [hunk 4]
```diff
@@ -133,10 +131,9 @@ struct execute_on_stream_nosync : execute_on_stream_nosync_base<execute_on_strea
 struct par_t : execution_policy<par_t>,
   thrust::detail::allocator_aware_execution_policy<
     execute_on_stream_base>
-#if THRUST_CPP_DIALECT >= 2011
 , thrust::detail::dependencies_aware_execution_policy<
     execute_on_stream_base>
-#endif
+#
```
Comment (ZelboK) [inline thrust/thrust/system/cuda/detail/par.h:136]: remove

File: thrust/internal/benchmark/bench.cu [hunk 5]
```diff
@@ -48,13 +44,8 @@
...
 #define PP_CAT(a, b) a ## b
 
-// We don't use THRUST_NOEXCEPT because it's new, and we want this benchmark to
-// be backwards-compatible to older versions of Thrust.
-#if THRUST_CPP_DIALECT >= 2011
-  #define NOEXCEPT noexcept
-#else
-  #define NOEXCEPT throw()
-#endif
+#define NOEXCEPT noexcept
```
Comment (miscco) [inline thrust/internal/benchmark/bench.cu:47]: Please just replace any instances of `NOEXCEPT` with `noexcept` and remove that macro

File: thrust/thrust/detail/alignment.h [hunk 6]
```diff
@@ -43,52 +41,17 @@ namespace detail
 /// inside of a `__declspec(align(#))` attribute. As a workaround, you can
 /// assign the result of \p THRUST_ALIGNOF to a variable and pass the variable
 /// as the argument to `__declspec(align(#))`.
-#if THRUST_CPP_DIALECT >= 2011
     #define THRUST_ALIGNOF(x) alignof(x)
```
Comment (miscco) [inline thrust/thrust/detail/alignment.h:44]: Ditto remove the macro completely

File: thrust/thrust/type_traits/is_trivially_relocatable.h [hunk 7]
```diff
@@ -260,17 +233,14 @@ template <typename T>
...
                 __is_trivially_copyable(T)
             #elif THRUST_HOST_COMPILER == THRUST_HOST_COMPILER_GCC && THRUST_GCC_VERSION >= 50000
                 std::is_trivially_copyable<T>::value
             #else
                 has_trivial_assign<T>::value
             #endif
-        #else
-            has_trivial_assign<T>::value
-        #endif
+
```
Comment (miscco) [inline thrust/thrust/type_traits/is_trivially_relocatable.h:243]:
```suggestion
```

File: thrust/thrust/zip_function.h [hunk 8]
```diff
@@ -10,8 +10,8 @@
 #include <thrust/detail/config.h>
 #include <thrust/detail/cpp11_required.h>
 #include <thrust/detail/modern_gcc_required.h>
-
-#if THRUST_CPP_DIALECT >= 2011 && !defined(THRUST_LEGACY_GCC)
+// remove?
```
Comment (miscco) [inline thrust/thrust/zip_function.h:13]: While we officially do not support those old gcc compilers anymore there is no benefit in intentionally breaking support right now, so I would leave that as is
```suggestion

```

**Review by miscco — COMMENTED** [review]
Great work 😸 
I had some minor comments about replacing now unneeded macros.

File: thrust/thrust/zip_function.h (hunk 8 above)
Comment (miscco) [reply thrust/thrust/zip_function.h:13]: In the sense as make it `#if !defined(THRUST_LEGACY_GCC)`

**Comment by miscco** [comment]
I believe you need to replace those macros like `THRUST_ALIGNOF`

**Comment by miscco** [comment]
I believe something went wrong here and you somehow added your sccache cache to the repository

File: thrust/testing/zip_function.cu [hunk 9]
```diff
@@ -1,6 +1,7 @@
 #include <thrust/detail/config.h>
 
-#if THRUST_CPP_DIALECT >= 2011 && !defined(THRUST_LEGACY_GCC)
+
+#if !defined(THRUST_LEGACY_GCC)
```
Comment (miscco) [inline thrust/testing/zip_function.cu:4]:
```suggestion
#if !defined(THRUST_LEGACY_GCC)
```

File: thrust/thrust/detail/allocator_aware_execution_policy.h [hunk 10]
```diff
@@ -26,9 +26,8 @@ _CCCL_IMPLICIT_SYSTEM_HEADER
 #include <thrust/detail/execute_with_allocator_fwd.h>
 #include <thrust/detail/alignment.h>
 
-#if THRUST_CPP_DIALECT >= 2011
-  #include <type_traits>
-#endif
+#include <type_traits>
+
```
Comment (miscco) [inline thrust/thrust/detail/allocator_aware_execution_policy.h:30]:
```suggestion
```

File: thrust/thrust/detail/config/cpp_compatibility.h [hunk 11]
```diff
@@ -60,25 +60,17 @@ _CCCL_IMPLICIT_SYSTEM_HEADER
...
-#  if THRUST_CPP_DIALECT >= 2011
 #    define THRUST_INLINE_CONSTANT                 static const __device__
 #    define THRUST_INLINE_INTEGRAL_MEMBER_CONSTANT static constexpr
-#  else
-#    define THRUST_INLINE_CONSTANT                 static const __device__
-#    define THRUST_INLINE_INTEGRAL_MEMBER_CONSTANT static const
-#  endif
+
 #else
 // FIXME: Add this when NVCC supports inline variables.
```
Comment (miscco) [inline thrust/thrust/detail/config/cpp_compatibility.h:67]: Filed #594 as a follow up for this

File: thrust/thrust/detail/tuple.inl [hunk 12]
```diff
@@ -0,0 +1,1000 @@
...
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#pragma once
+
+#include <thrust/detail/config.h>
```
Comment (miscco) [inline thrust/thrust/detail/tuple.inl:19]: That file is a merge conflict from the recent replacement of `thrust::tuple` with `cuda::std::tuple`

Please remove the file

File: thrust/thrust/mr/allocator.h [hunk 13]
```diff
@@ -243,9 +227,8 @@ class stateless_resource_allocator : public thrust::mr::allocator<T, Upstream>
     stateless_resource_allocator(const stateless_resource_allocator<U, Upstream> & other)
         : base(other) {}
 
-#if THRUST_CPP_DIALECT >= 2011
     stateless_resource_allocator & operator=(const stateless_resource_allocator &) = default;
-#endif
+
```
Comment (miscco) [inline thrust/thrust/mr/allocator.h:231]:
```suggestion
```

File: thrust/thrust/mr/disjoint_pool.h [hunk 14]
```diff
@@ -257,9 +257,7 @@ class disjoint_unsynchronized_pool_resource final
         {
         }
 
-#if THRUST_CPP_DIALECT >= 2011
         pool & operator=(const pool &) = default;
-#endif
 
         __host__
         ~pool() {}
```
Comment (miscco) [inline thrust/thrust/mr/disjoint_pool.h:263]: Filed #595 to get rid of all those user-defined destructors

**Review by miscco — COMMENTED** [review]
Mostly whitespace issues and a merge conflict. 

Thanks a lot for cleaning this up 🥇

File: thrust/thrust/detail/allocator_aware_execution_policy.h [hunk 15]
```diff
@@ -27,7 +27,7 @@ _CCCL_IMPLICIT_SYSTEM_HEADER
 #include <thrust/detail/alignment.h>
 
 #include <type_traits>
-
+#include <type_traits>
```
Comment (miscco) [inline thrust/thrust/detail/allocator_aware_execution_policy.h:30]: Urgh sorry, that seems to have been an issue with the UI, I accidentally added the header
```suggestion
```

**Review by miscco — APPROVED** [review]

**Comment by miscco** [comment]
@ZelboK thanks a lot for working on this. This is definitely ready to merge. That said, we are currently in the process of releasing 2.3.x and we would like to wait for that to be done, as this PR could potentially create a lot of merge conflicts

File: thrust/testing/complex.cu [hunk 16]
```diff
@@ -708,22 +705,3 @@ struct TestComplexStdComplexDeviceInterop
...
-
-  void operator()()
-  {
-    const thrust::complex<T> input(42.0, 1337.0);
-    const user_complex result = thrust::exp(input);
-    (void)result;
-  }
-};
-SimpleUnitTest<TestComplexExplicitConstruction, FloatingPointTypes>
-  TestComplexExplicitConstructionInstance;
```
Comment (ZelboK) [inline thrust/testing/complex.cu:729]: (comment for self) 
why did i remove this?

File: thrust/thrust/detail/alignment.h [hunk 17]
```diff
@@ -105,8 +56,8 @@ namespace detail
 template <std::size_t Align>
 struct aligned_type;
 
-#if _CCCL_STD_VER >= 2011                                                     \
-  && defined(_CCCL_COMPILER_GCC) && (THRUST_GCC_VERSION >= 40800)
+#if (THRUST_HOST_COMPILER == THRUST_HOST_COMPILER_GCC) \
```
Comment (ZelboK) [inline thrust/thrust/detail/alignment.h:59]: (Comment for self)
Fix

File: thrust/thrust/detail/allocator/allocator_traits.h [hunk 18]
```diff
@@ -401,10 +333,8 @@ template<typename Alloc>
   template<typename T, typename Arg1>
   inline _CCCL_HOST_DEVICE static void construct(allocator_type &a, T *p, const Arg1 &arg1);
 
-#if _CCCL_STD_VER >= 2011
   template<typename T, typename... Args>
-  inline _CCCL_HOST_DEVICE static void construct(allocator_type &a, T *p, Args&&... args);
-#endif
+  inline __host__ __device__ static void construct(allocator_type &a, T *p, Args&&... args);
```
Comment (ZelboK) [inline thrust/thrust/detail/allocator/allocator_traits.h:337]: (Comment for self)
Fix.

File: thrust/thrust/detail/config/cpp_compatibility.h [hunk 19]
```diff
@@ -62,25 +62,17 @@
 //#  if   _CCCL_STD_VER >= 2017
 //#    define THRUST_INLINE_CONSTANT                 inline constexpr
 //#    define THRUST_INLINE_INTEGRAL_MEMBER_CONSTANT inline constexpr
-#  if _CCCL_STD_VER >= 2011
-#    define THRUST_INLINE_CONSTANT                 static const _CCCL_DEVICE
+#    define THRUST_INLINE_CONSTANT                 static const __device__
```
Comment (ZelboK) [inline thrust/thrust/detail/config/cpp_compatibility.h:65]: Fix

File: thrust/thrust/zip_function.h [hunk 20]
```diff
@@ -18,8 +18,7 @@
 #endif // no system header
 #include <thrust/detail/cpp11_required.h>
 #include <thrust/detail/modern_gcc_required.h>
-
-#if _CCCL_STD_VER >= 2011 && !defined(THRUST_LEGACY_GCC)
+#if !defined(THRUST_LEGACY_GCC)
```
Comment (ZelboK) [inline thrust/thrust/zip_function.h:21]: Correct me if im wrong but is there another macro for this one?

**Comment by ZelboK** [comment]
@jrhemstad I looked over teh PR, I think this should be ready. Could use another set of 👀

**Review by miscco — APPROVED** [review]

File: thrust/thrust/zip_function.h (hunk 20 above)
Comment (miscco) [reply thrust/thrust/zip_function.h:21]: Nope, but that is fine

**Comment by ZelboK** [comment]
@miscco Do I need to do anything from my end? Should be good to go?

**Comment by miscco** [comment]
Thanks a lot for keeping the codebase up to date 🎉

**Pull request merged**